sw/otbn/crypto/field25519.s

```diff
@@ -0,0 +1,350 @@
+/* Copyright lowRISC contributors. */
+/* Licensed under the Apache License, Version 2.0, see LICENSE for details. */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+/**
+ * This library contains arithmetic for the finite field modulo the prime
+ * p = 2^255-19, which is used for both the X25519 key exchange and the Ed25519
+ * signature scheme.
+ */
+
+
+/**
+ * Fully reduce a 510-bit number modulo p.
+ *
+ * Returns c = a mod p.
+ *
+ * Uses Solinas/pseudo-Mersenne reduction, which is based on the observation
+ * that if a large number x is split into two so that the lower 255 bits are x0
+ * and all bits above 255 are x1, then:
+ *
+ *   x = x0 + (2^255 * x1) \equiv (x0 + (19 * x1)) (mod p)
+ *
+ * If x is large, then x0 + 19 * x1 will be much smaller than x, because 19 <<
+ * 2^255. This step can be repeated as necessary until a conditional
+ * subtraction is enough to fully reduce.
+ *
+ * This routine runs in constant time.
+ *
+ * Flags: Flags have no meaning beyond the scope of this subroutine.
+ *
+ * @param[in]  w19: constant, w19 = 19
+ * @param[in]  [w21:w20]: a, number to be reduced (a < 2^510)
+ * @param[in]  MOD: p, modulus = 2^255 - 19
+ * @param[in]  w31: all-zero
+ * @param[out] w22: c, result = a mod p
+ *
+ * clobbered registers: w18, w20 to w22
+ * clobbered flag groups: FG0
+ */
+fe_reduce:
+  /* First Solinas step (reducing modulo 2p = 2^256-38). */
+
+  /* Multiply the high bits by 38 (fits in 64bx256b multiply). Note that the
+     last multiplication result is zero; it exists only to write back the
+     accumulator from previous multiplies.
+     w18,w22 <= 19 * (2 * w21) = 38 * a[509:257] */
+  bn.add                w18, w21, w21
+  bn.mulqacc.z          w19.0, w18.0, 0
+  bn.mulqacc.so  w22.L, w19.0, w18.1, 64
+  bn.mulqacc            w19.0, w18.2, 0
+  bn.mulqacc.so  w22.U, w19.0, w18.3, 64
+  bn.mulqacc.wo  w18,   w19.0, w31.0, 0
+
+  /* Add to low bits.
+     [w21:w20] <= w20 + [w18,w22] = a[256:0] + (38 * a[509:257]) = r1 */
+  bn.add  w20, w20, w22
+  bn.addc w21, w31, w18
+
+  /* First Solinas step is complete. At this point, the new intermediate result
+     r1 is at most 261 bits, because:
+                        a[509:257] = 254 bits
+                   38 * a[509:257] = 260 bits
+        a[256:0] + 38 * a[509:257] = 261 bits */
+
+  /* Begin second Solinas step (reducing by p = 2^255 - 19 this time). */
+
+  /* Extract the high 6 bits.
+     w21 <= [w21:w20] >> 255 = r1[260:256] */
+  bn.rshi w21, w21, w20 >> 255
+
+  /* Extract the low 255 bits.
+     w20 <= r1[255:0] */
+  bn.rshi w20, w20, w31 >> 255
+  bn.rshi w20, w31, w20 >> 1
```

Comment:
Yes, it's clearing the MSB. And yes, I opted against the mask because in order to save an instruction, we'd have to use it as a constant input to this function and have callers precompute it, which could be worth it for performance but also means an extra register used and complexity in the callers. But I should record that possibility in a comment here at least, so if we're looking for performance wins later we can try it.